server/src/main/java/org/opensearch/index/engine/NRTReplicationReaderManager.java

```diff
@@ -98,12 +98,12 @@ protected OpenSearchDirectoryReader refreshIfNeeded(OpenSearchDirectoryReader re
      * @param infos {@link SegmentInfos} infos
      * @throws IOException - When Refresh fails with an IOException.
      */
-    public synchronized void updateSegments(SegmentInfos infos) throws IOException {
+    public void updateSegments(SegmentInfos infos) throws IOException {
```

Comment:
Thanks @mch2 for this quick fix. I have couple of questions. 
1. Why do we want the guarantee that update thread performs a refresh. I assume the refresh thread would be doing the same ?
2. How do we know the issue is fixed with your change ?

Reply:
1. The refresh lock could still be held by a different thread after it has returned null (no refresh )from refreshIfNeeded.  So we need to ensure a refresh has executed if the update thread does not immediately get the lock.
2. I've added a unit test that reproduces this scenario, it is fixed with the update.

Reply:
Thanks @mch2 for the prompt response. :)